python/lsst/daf/butler/core/datastore.py

```diff
@@ -228,6 +246,45 @@ def put(self, inMemoryDataset, datasetRef):
         """
         raise NotImplementedError("Must be implemented by subclass")
 
+    def ingest(self, path, ref, formatter=None, transfer=None):
```

Comment:
Yeah, I've been thinking about this.  From a pure OO perspective, it'd be very natural to have an intermediate (but still abstract) base class `FileDatastore` that knows about formatters and files and ingest.  All concrete `Datastores` that work with formatters and files and can do ingest would then inherit from that instead of directly from `Datastore` itself, and that might include, say, an S3 Datastore that doesn't have files internally certailny knows how to work with them.

Anyhow, we'd then be able to define a  bunch of operations that only work when you have a `FileDatastore`. Ingest is the obvious operation that's relevant here, but I wonder if this could even be extended to the problem of whether you can ask for a checksum or size (i.e. "yes, but only if you have a `FileDatastore`").

Unfortunately, I don't think the derived-class-interface approach works with composition (a `ChainedDatastore` can't inherit from `FileDatastore` if and only if it forwards to a `FileDatastore`).  I think that means we can't use the Python type system to express "is a file datastore", but the "file datastore" interface concept may still be one worth preserving.  That's the direction I was trying to go in here, but it's not really fully thought out and I do agree that we should do that thinking before letting paths and formatters spread too much further into generic datastores.